doc/articles/concepts/source-generation.md

```diff
@@ -0,0 +1,51 @@
+
+Uno can however, generate code. 
+
+With Roslyn, Microsoft open-sourced the C# compiler, but they also exposed a powerful API for code analysis. Roslyn provides a easy to access all the syntactic and semantic information that the compiler possesses. Uno created a [source generator](https://github.com/nventive/Uno.SourceGeneration) that leverages this power for code generation and like the Uno platform, it's free and open-source.
+
+Inside of the [Uno.SourceGeneration](https://github.com/nventive/Uno.SourceGeneration) you'll find a msbuild task that allows you to easily add generated code based on Roslyn's analysis of your solution. This might be partial class definitions which augment existing types, or it might be entirely new classes. 
+
+In Uno, this used by the [DependencyObjectGenerator](https://github.com/nventive/Uno/blob/master/src/SourceGenerators/Uno.UI.SourceGenerators/DependencyObject/DependencyObjectGenerator.cs) class. This generator looks for every class in the solution that implements the `DependencyObject` interface. For each such class, it automatically generates the methods and properties of `DependencyObject`. 
+
+Since the generator has a full set of semantic information from Roslyn, it can do this in a smart way. For instance, if it detects that the class is a view type, it adds methods to update binding information when the view is loaded or unloaded. 
+
+Here's a [small snippet](https://github.com/nventive/Uno/blob/74ba91756c446107e7394e0423527de273154f5d/src/SourceGenerators/Uno.UI.SourceGenerators/DependencyObject/DependencyObjectGenerator.cs#L218-L250) of code from `DependencyObjectGenerator` which demonstrates this. In the snippet there is a [INamedTypeSymbol](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.inamedtypesymbol?view=roslyn-dotnet), an object from Roslyn that encapsulates information about a type and the following logic:
+
+- A determination that `typeSymbol` implements `DependencyObject`.
+- A check if it's an Android `View` and, if so override the loaded method. 
+- A check to ensure the type doesn't _already_ override the same method so that the generator doesn't accidentally generate code that clashes with authored code and causes a compiler error. 
+
+All this goes on under the hood without user intervention, whenever your app compiles. The end result is that `DependencyObject` can be used almost exactly the same way with Uno as with UWP, even though it's an interface and not a class.
+
+# Generation of native constructors
+
+tba
```

Comment:
💭

Reply:
Because views in Uno inherit directly from native views on Android/iOS, they need to have special constructors that are called under-the-hood by Xamarin (https://docs.microsoft.com/en-us/xamarin/android/platform/java-integration/working-with-jni#binding-constructors). 

This is tedious to write by hand, and particularly painful when porting existing UWP code, so Uno adds it automatically with the source generator.   (Checking first that it's not already there https://github.com/unoplatform/uno/blob/master/src/SourceGenerators/Uno.UI.SourceGenerators/NativeCtor/NativeCtorsGenerator.cs#L126)